vortex-ffi/src/array_builder.rs

```diff
@@ -0,0 +1,468 @@
+            pub unsafe extern "C-unwind" fn [<vx_array_primitive_new_ $ptype>](
+                data: *const $rust_type,
+                len: usize,
+                validity: *const bool,
+                error_out: *mut *mut vx_error,
+            ) -> *const vx_array {
+                try_or_default(error_out, || {
+                    if data.is_null() {
+                        return Err(vortex_err!("data pointer is null"));
+                    }
+
+                    let data_slice = unsafe { slice::from_raw_parts(data, len) };
+                    let mut buffer = BufferMut::<$rust_type>::with_capacity(len);
+                    buffer.extend_from_slice(data_slice);
+
+                    let validity = if validity.is_null() {
+                        Validity::NonNullable
+                    } else {
+                        let validity_slice = unsafe { slice::from_raw_parts(validity, len) };
+                        Validity::from_iter(validity_slice.iter().copied())
+                    };
+
+                    let array = PrimitiveArray::new(buffer.freeze(), validity).into_array();
+                    Ok(vx_array::new(array))
+                })
+            }
+        }
+    };
+}
+
+ffi_primitive_array_new!(u8, u8);
+ffi_primitive_array_new!(u16, u16);
+ffi_primitive_array_new!(u32, u32);
+ffi_primitive_array_new!(u64, u64);
+ffi_primitive_array_new!(i8, i8);
+ffi_primitive_array_new!(i16, i16);
+ffi_primitive_array_new!(i32, i32);
+ffi_primitive_array_new!(i64, i64);
+ffi_primitive_array_new!(f16, f16);
+ffi_primitive_array_new!(f32, f32);
+ffi_primitive_array_new!(f64, f64);
+
+// =============================================================================
+// Bool Array Builder
+// =============================================================================
+
+/// Create a new boolean array from raw data.
+///
+/// The `data` pointer must point to a valid array of `len` booleans.
+/// The `validity` pointer, if not null, must point to a valid array of `len` booleans.
+/// If `validity` is null, the array is assumed to have no null values.
+///
+/// Returns a new array, or null on error.
+#[unsafe(no_mangle)]
+pub unsafe extern "C-unwind" fn vx_array_bool_new(
+    data: *const bool,
+    len: usize,
+    validity: *const bool,
+    error_out: *mut *mut vx_error,
+) -> *const vx_array {
+    try_or_default(error_out, || {
+        if data.is_null() {
+            return Err(vortex_err!("data pointer is null"));
+        }
+
+        let data_slice = unsafe { slice::from_raw_parts(data, len) };
+
+        let nullability = if validity.is_null() {
+            Nullability::NonNullable
+        } else {
+            Nullability::Nullable
+        };
+
+        let mut builder = BoolBuilder::with_capacity(nullability, len);
+
+        for (i, &value) in data_slice.iter().enumerate() {
+            if !validity.is_null() {
+                let validity_slice = unsafe { slice::from_raw_parts(validity, len) };
```

Comment:
Is this free? It seems at least for legibility's sake, we should lift it maybe to line 101?